File: 2025-1/sesion-de-clase-23/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/logistica/almacen/ProductoDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.logistica.almacen;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.logistica.almacen.IProductoDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAOImpl;
import pe.edu.pucp.inf30.softprog.model.logistica.almacen.Producto;

/**
 *
 * @author eric
 */
public class ProductoDAOImpl extends BaseDAOImpl<Producto> implements IProductoDAO {
    @Override
    protected CallableStatement comandoInsertar(Connection conn, Producto modelo) throws SQLException {
        String sql = "{CALL insertarProducto(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_unidadMedida", modelo.getUnidadMedida());
        cmd.setDouble("p_precio", modelo.getPrecio());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }

    @Override
    protected CallableStatement comandoModificar(Connection conn, Producto modelo) throws SQLException {
        String sql = "{CALL modificarProducto(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_unidadMedida", modelo.getUnidadMedida());
        cmd.setDouble("p_precio", modelo.getPrecio());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        return cmd;
    }

    @Override
    protected CallableStatement comandoEliminar(Connection conn, int id) throws SQLException {
        String sql = "{CALL eliminarProducto(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoBuscar(Connection conn, int id) throws SQLException {
        String sql = "{CALL buscarProductoPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoListar(Connection conn) throws SQLException {
        String sql = "{CALL listarProductos()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }

    @Override
    protected Producto mapearModelo(ResultSet rs) throws SQLException {
        Producto producto = new Producto();
        producto.setId(rs.getInt("id"));
        producto.setNombre(rs.getString("nombre"));
        producto.setUnidadMedida(rs.getString("unidadMedida"));
        producto.setPrecio(rs.getDouble("precio"));
        producto.setActivo(rs.getBoolean("activo"));
        return producto;
    }
}

File: 2025-1/sesion-de-clase-23/SoftProgNet/SoftProgPersistencia/DAOImpl/Logistica/Almacen/ProductoDAOImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PUCP.Edu.Pe.SoftProg.Modelo.Clientes;
using PUCP.Edu.Pe.SoftProg.Modelo.Logistica.Almacen;
using PUCP.Edu.Pe.SoftProg.Modelo.RRHH;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO.Logistica.Almacen;

namespace PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl.Logistica.Almacen {
    public class ProductoDAOImpl : BaseDAOImpl<Producto>, IProductoDAO {
        protected override DbCommand CommandoInsertar(DbConnection conn, Producto producto) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarProducto";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, producto.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_unidadMedida", DbType.String, producto.UnidadMedida);
            this.AgregarParametroEntrada(cmd, "@p_precio", DbType.Double, producto.Precio);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, producto.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand CommandoInsertar(DbConnection conn, DbTransaction transaccion, Producto producto) {
            DbCommand cmd = this.CommandoInsertar(conn, producto);
            cmd.Transaction = transaccion;

            return cmd;
        }

        protected override DbCommand CommandoModificar(DbConnection conn, Producto producto) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarProducto";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, producto.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_unidadMedida", DbType.String, producto.UnidadMedida);
            this.AgregarParametroEntrada(cmd, "@p_precio", DbType.Double, producto.Precio);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, producto.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, producto.Id);

            return cmd;
        }

        protected override DbCommand CommandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarProducto";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoBuscar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarProductoPorId";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoListar(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarProductos";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            return cmd;
        }

        protected override Producto mapearModelo(DbDataReader reader) {
            return new Producto {
                Id = Convert.ToInt32(reader["id"]),
                Nombre = Convert.ToString(reader["nombre"]),
                UnidadMedida = Convert.ToString(reader["unidadMedida"]), 
                Precio = Convert.ToDouble(reader["precio"]), 
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }
    }
}
